### PR TITLE
[AMORO-2008]: Refactor the optimizer module to separate distributing pacakges for different engines.

### LINKED ISSUE #2008: [Improvement]: Separate `local-optimizer` and `flink-optimizer` into different modules and package them independently.
### Search before asking

- [X] I have searched in the [issues](https://github.com/NetEase/amoro/issues?q=is%3Aissue) and found no similar issues.

### What would you like to be improved?

Currently, the Optimizing job is a fat-jar that is applicable for both standalone mode and Flink job mode. This is not conducive to future expansion of more deployment methods and may cause conflicts during dependency packaging. 

A typical example is that when deploying independently, Hadoop-related dependencies need to be packaged, but in Flink job mode, there may be conflicts with the Hadoop dependencies that come with the Flink engine.

Therefore, independent packaging for each type is a better choice.

### How should we improve?

1. The optimizer module will be split into three sub-modules: optimizer/common, optimizer/jvm, and optimizer/flink.
2. The optimizer/jvm sub-module, as a fatjar, will include various dependencies such as Hadoop and AWS.
3. The optimizer/flink sub-module will only be used as a Flink job and will not include lower-level dependencies.
4. The optimizer/jvm and optimizer/flink sub-modules will depend on the optimizer/common sub-module. The logic related to optimizer/task execution will be implemented in this module.

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Subtasks

_No response_

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by baiyangtx**
## Why are the changes needed?
 

Close #2008 .

## Brief change log

- The optimizer module has been reorganized and split into `ams/optimizer-container`, `ams/optimizer-job/job-common`, `ams/optimizer-job/flink-job`, and `ams/optimizer-job/standalone-job`. 
- Changes have been made to the configuration files such as `config.yaml`, helm `configmap`, etc., to update the package name of localOptimizerContainer. 
- The dist directory has been modified, and the `plugin/optimize/OptimizeJob.jar` has been removed and replaced with `plugin/optimize/flink/optimize-job.jar`, which is only used for the Flink optimizer.

## How was this patch tested?

- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [X] Run test locally before making a pull request

## Documentation

- Does this pull request introduce a new feature? (no)

### CODE REVIEW

File: ams/optimizer-container/pom.xml [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>amoro-optimizer-container</artifactId>
+    <name>Amoro Project Optimizer Container</name>
+    <url>https://amoro.netease.com</url>
+
+    <properties>
+        <fastjson.version>1.2.58</fastjson.version>
+        <avro.version>1.10.1</avro.version>
+        <snakeyaml.version>1.30</snakeyaml.version>
```
Comment (shidayang) [inline ams/optimizer-container/pom.xml:37]: According to the current specification, all dependencies must be added to the dependencyManagement.

**Comment by zhoujinsong** [comment]
@baiyangtx Thanks a lot for your work!

`Optimizer job` seems to be a new concept for Amoro developers and users. Here, I think using `Optimizer` directly seems accurate enough. So I suggested to name the module to :

- ams/optimizer-container
- ams/optimizer/common
- ams/optimizer/standalone-optimizer
- ams/optimizer/flink-optimizer

How do you think?

**Comment by baiyangtx** [comment]
> @baiyangtx Thanks a lot for your work!
> 
> `Optimizer job` seems to be a new concept for Amoro developers and users. Here, I think using `Optimizer` directly seems good enough. So I suggested to name the module to :
> 
>     * ams/optimizer-container
> 
>     * ams/optimizer/common
> 
>     * ams/optimizer/standalone-optimizer
> 
>     * ams/optimizer/flink-optimizer
> 
> 
> How do you think?

I agree with it. I will fix it.

File: ams/optimizer-container/pom.xml (hunk 1 above)
Comment (zhoujinsong) [reply ams/optimizer-container/pom.xml:37]: Yes, #2030 has move these properties to parent pom.

File: ams/optimizer/common/src/main/java/com/netease/arctic/optimizer/job/common/Optimizer.java [hunk 2]
```diff
@@ -1,4 +1,4 @@
-package com.netease.arctic.optimizer;
+package com.netease.arctic.optimizer.job.common;
```
Comment (zhoujinsong) [inline ams/optimizer/common/src/main/java/com/netease/arctic/optimizer/job/common/Optimizer.java:1]: Package `com.netease.arctic.optimizer.job.common` seems to be too long for me, `com.netease.arctic.optimizer.job.common` is good enough for me.

File: ams/optimizer/flink-optimizer/pom.xml [hunk 3]
```diff
@@ -24,11 +24,12 @@
...
         <groupId>com.netease.amoro</groupId>
         <version>0.6.0-SNAPSHOT</version>
+        <relativePath>../../pom.xml</relativePath>
     </parent>
     <modelVersion>4.0.0</modelVersion>
 
-    <artifactId>amoro-ams-optimizer</artifactId>
-    <name>Amoro Project Optimizer</name>
+    <artifactId>flink-optimizer</artifactId>
+    <name>Amoro Project Flink Optimizer Job</name>
```
Comment (zhoujinsong) [inline ams/optimizer/flink-optimizer/pom.xml:32]:
```suggestion
    <name>Amoro Project Flink Optimizer</name>
```

File: ams/optimizer/standalone-optimizer/pom.xml [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    <parent>
+        <artifactId>amoro-ams</artifactId>
+        <groupId>com.netease.amoro</groupId>
+        <version>0.6.0-SNAPSHOT</version>
+        <relativePath>../../pom.xml</relativePath>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>standalone-optimizer</artifactId>
+    <name>Amoro Project Standalone Optimizer Job</name>
```
Comment (zhoujinsong) [inline ams/optimizer/standalone-optimizer/pom.xml:32]:
```suggestion
    <name>Amoro Project Standalone Optimizer</name>
```

**Review by zhoujinsong — COMMENTED** [review]
@baiyangtx  I left some comments.
BTW, you may need to modify the `managing-optimizer.md` docs too.

File: ams/optimizer/standalone-optimizer/pom.xml (hunk 4 above)
Comment (baiyangtx) [reply ams/optimizer/standalone-optimizer/pom.xml:32]: fixed

File: ams/optimizer/flink-optimizer/pom.xml (hunk 3 above)
Comment (baiyangtx) [reply ams/optimizer/flink-optimizer/pom.xml:32]: fixed

File: ams/optimizer/common/pom.xml [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    <parent>
+        <artifactId>amoro-ams</artifactId>
+        <groupId>com.netease.amoro</groupId>
+        <version>0.6.0-SNAPSHOT</version>
+        <relativePath>../../pom.xml</relativePath>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>optimizer-common</artifactId>
+    <name>Amoro Project Optimize Job Common</name>
```
Comment (zhoujinsong) [inline ams/optimizer/common/pom.xml:32]:
```suggestion
    <name>Amoro Project Optimizer Common</name>
```

File: ams/optimizer-container/pom.xml [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    <name>Amoro Project Optimizer Container</name>
+    <url>https://amoro.netease.com</url>
+
+    <dependencies>
+        <dependency>
+            <groupId>com.netease.amoro</groupId>
+            <artifactId>optimizer-common</artifactId>
+            <version>${project.version}</version>
+        </dependency>
+        <dependency>
```
Comment (zhoujinsong) [inline ams/optimizer-container/pom.xml:40]: Perhaps line breaks can be added here for better formatting?

File: ams/optimizer-container/pom.xml [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+            <groupId>org.yaml</groupId>
+            <artifactId>snakeyaml</artifactId>
+        </dependency>
+
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <scope>test</scope>
```
Comment (zhoujinsong) [inline ams/optimizer-container/pom.xml:48]: scope can be ommited.

File: ams/optimizer/common/pom.xml [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+        <dependency>
+            <groupId>args4j</groupId>
+            <artifactId>args4j</artifactId>
+            <version>${args4j.version}</version>
+        </dependency>
+
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <scope>test</scope>
```
Comment (zhoujinsong) [inline ams/optimizer/common/pom.xml:55]: Scope can be omited.

File: ams/optimizer/common/pom.xml [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+        <dependency>
+            <groupId>com.netease.amoro</groupId>
+            <artifactId>amoro-ams-api</artifactId>
+            <version>${project.version}</version>
+            <classifier>tests</classifier>
+            <scope>test</scope>
+        </dependency>
+    </dependencies>
+
+    <build>
```
Comment (zhoujinsong) [inline ams/optimizer/common/pom.xml:75]: Do we need to declare this plugin here and all other modules?

File: ams/optimizer/common/pom.xml (hunk 5 above)
Comment (baiyangtx) [reply ams/optimizer/common/pom.xml]: fixed.

File: ams/optimizer-container/pom.xml (hunk 6 above)
Comment (baiyangtx) [reply ams/optimizer-container/pom.xml:41]: fixed

File: ams/optimizer-container/pom.xml (hunk 7 above)
Comment (baiyangtx) [reply ams/optimizer-container/pom.xml]: fixed

File: ams/optimizer/common/pom.xml (hunk 8 above)
Comment (baiyangtx) [reply ams/optimizer/common/pom.xml]: fixed

File: ams/optimizer/common/pom.xml (hunk 9 above)
Comment (baiyangtx) [reply ams/optimizer/common/pom.xml]: fixed

**Pull request merged**